airflow/jobs/local_task_job.py

```diff
@@ -95,6 +95,14 @@ def signal_handler(signum, frame):
                 # Monitor the task to see if it's done
                 return_code = self.task_runner.return_code()
                 if return_code is not None:
+                    if return_code != 0:
+                        self.task_instance.refresh_from_db()
+                        # there is one case we should not treat non zero return
+                        # code as failed: the job has been killed externally.
+                        if (not self.terminating) or self.task_instance.state == State.FAILED:
```

Comment:
How does `state="failed"` relate to job been killed externally?

I'm also not that happy about having very similar conditions here and inside on_failure. And again: Why throw an exception only to have it caught later, why not just call `self.on_failure`?

Reply:
1. "self.terminating" means job been killed externally but not "state="failed". It's set in heartbeat_callback:L137. Basically it means whenever "ti.state != State.RUNNING", it's terminating.

There are two cases this will happen
a. explicitly calling terminate of a StandardTaskRunner, then the return code is -9 and the task_instance.state is not failed. In such case, we should not treat non zero exit code as failure.
b. the task instance is explicitly set as failed, in this case, we should treat zero exit code as failure.

2. As mentioned above, "job failure" is handled in base_job.py:run:L230. If exception is not thrown here, job state will not be marked as failure